stan/math/opencl/kernel_generator/transpose.hpp

```diff
@@ -0,0 +1,131 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_TRANSPOSE_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_TRANSPOSE_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <algorithm>
+#include <string>
+#include <set>
+#include <type_traits>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents a transpose in kernel generator expressions.
+ *
+ * Warning: transposing this expression is not supported!
+ * @tparam Derived derived type
+ * @tparam T_a type of first argument
+ * @tparam T_b type of second argument
+ */
+template <typename T>
+class transpose_
+    : public operation_cl<transpose_<T>,
+                          typename std::remove_reference_t<T>::Scalar, T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl<transpose_<T>, Scalar, T>;
+  using base::var_name;
+
+ protected:
+  using base::arguments_;
+
+ public:
+  /**
+   * Constructor
+   * @param a expression to transpose
+   */
+  explicit transpose_(T&& a) : base(std::forward<T>(a)) {}
+
+  /**
+   * Creates a deep copy of this expression.
+   * @return copy of \c *this
+   */
+  inline transpose_<std::remove_reference_t<T>> deep_copy() {
+    return transpose_<std::remove_reference_t<T>>{
+        std::get<0>(arguments_).deep_copy()};
+  }
```

Comment:
This is kinda weird to me. Let's say we call 
```cpp
double b = 10;
transpose_<double> foo(b);
``` 
(or insert any non ref type). Then the constructor here is going to be 

```cpp
explicit transpose_(double&& a) : base(std::forward<double>(a)) {}
```

Where `std::forward<double>(a)` here is actually going to give back an rvalue reference since forward with an lvalue as the template and rvalue as the type still produces an rvalue (see godbolt below). 

I think you need to have a separate template here
```cpp
template <typename OpT, require_same_t<T, OpT>* = nullptr>
explicit transpose_(OpT&& a) : base(std::forward<OpT>(a)) {}
```

The godbolt here shows this effect.

https://godbolt.org/z/pWNzSj

Reply:
Having rvalue if rvalue is passed in here is fine. Your example does not compile because you have wrong template argument in `S<int> a(b);` - this should be `S<int&> a(b);`

Reply:
Also in your example in the comment you have an error - the constructor for 
```
double b = 10;
transpose_<double> foo(b);
```
would be:
```
explicit transpose_(double& a) : base(std::forward<double&>(a)) {}
```